src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -91,8 +108,92 @@ public void failed(Throwable exc, Void v) {
+        if (DiskBasedCacheUtility.wouldExceedCacheSize(
+                        mTotalSize + entry.data.length, mMaxCacheSizeInBytes)
+                && DiskBasedCacheUtility.isDataTooLarge(entry.data.length, mMaxCacheSizeInBytes)) {
+            return;
+        }
+
+        final File file = DiskBasedCacheUtility.getFileForKey(key, mRootDirectorySupplier);
+        Path path = Paths.get(file.getPath());
+
+        try {
+            final AsynchronousFileChannel afc =
+                    AsynchronousFileChannel.open(
+                            path, StandardOpenOption.WRITE, StandardOpenOption.CREATE);
+            final CacheHeader header = new CacheHeader(key, entry);
+            int headerSize = header.getHeaderSize();
+            int size = entry.data.length + headerSize;
+            ByteBuffer buffer = ByteBuffer.allocate(size);
+            header.writeHeader(buffer);
+            buffer.put(entry.data);
+            buffer.flip();
+            afc.write(
+                    /* source= */ buffer,
+                    /* position= */ 0,
+                    /* attachment= */ null,
+                    new CompletionHandler<Integer, Void>() {
+                        @Override
+                        public void completed(Integer resultLen, Void ignore) {
+                            header.size = resultLen;
+                            mTotalSize =
+                                    DiskBasedCacheUtility.putEntry(
+                                            key, header, mTotalSize, mEntries);
+                            mTotalSize =
+                                    DiskBasedCacheUtility.pruneIfNeeded(
+                                            mTotalSize,
+                                            mMaxCacheSizeInBytes,
+                                            mEntries,
+                                            mRootDirectorySupplier);
+                            try {
+                                afc.close();
+                            } catch (IOException e) {
+                                VolleyLog.e(e, "Failed to close channel");
+                            }
+                            callback.onPutComplete();
+                        }
+
+                        @Override
+                        public void failed(Throwable throwable, Void aVoid) {
+                            VolleyLog.e(
+                                    throwable, "Failed to read file %s", file.getAbsolutePath());
+                            callback.onPutComplete();
+                            try {
+                                afc.close();
+                            } catch (IOException e) {
+                                VolleyLog.e(e, "Failed to close channel");
+                            }
+                        }
+                    });
+        } catch (IOException e) {
+            boolean deleted = file.delete();
```

Comment:
Same here, should we close the channel if it was open?

Reply:
It seems like we should try to close it before trying to delete the file. I would imagine that deleting a file and then calling close() could cause the file to reappear, but not sure.

Reply:
if the close fails, would we still try to delete the file?